### PR TITLE
Convert `typedef` statements to C++11's `using`

### PR DESCRIPTION
**Pull request by Calinou**
This provides a more readable syntax that is recommended since C++11.

### CODE REVIEW

**Comment by timothyqiu** [comment]
> Note that I didn't convert statements such as these ones yet, since I don't know the syntax changes required to make them compile

I think you just have to remove the type name:

```c++
// typedef BroadPhase2DSW *(*CreateFunction)();
using CreateFunction = BroadPhase2DSW *(*)();

// typedef void *(*PairCallback)(CollisionObject2DSW *A, int p_subindex_A, CollisionObject2DSW *B, int p_subindex_B, void *p_userdata); 
using PairCallback = void *(*)(CollisionObject2DSW *A, int p_subindex_A, CollisionObject2DSW *B, int p_subindex_B, void *p_userdata);
```

File: core/extension/gdnative_interface.h [hunk 1]
```diff
@@ -127,14 +127,14 @@ typedef enum {
...
-typedef uint8_t GDNativeBool;
+using GDNativeVariantPtr = void *;
+using GDNativeStringNamePtr = void *;
+using GDNativeStringPtr = void *;
+using GDNativeObjectPtr = void *;
+using GDNativeTypePtr = void *;
+using GDNativeMethodBindPtr = void *;
+using GDNativeInt = int64_t;
+using GDNativeBool = uint8_t;
 typedef uint64_t GDObjectInstanceID;
```
Comment (merumelu) [inline core/extension/gdnative_interface.h:138]: Forgot this one?

**Comment by Calinou** [comment]
So far, I converted most `typedef`s except `typedef struct` and `typedef enum`.

File: modules/mono/mono_gd/gd_mono.h [hunk 2]
```diff
@@ -109,7 +109,7 @@ class GDMono {
 	LoadedApiAssembly core_api_assembly;
 	LoadedApiAssembly editor_api_assembly;
 
-	typedef bool (*CoreApiAssemblyLoadedCallback)();
+	using (*CoreApiAssemblyLoadedCallback)() = bool;
```
Comment (neikeq) [inline modules/mono/mono_gd/gd_mono.h:112]: This seems wrong. Looking at the other changes I think this should be like this?:

```suggestion
	using CoreApiAssemblyLoadedCallback = bool (*)();
```

**Comment by reduz** [comment]
This looks very good, I think it may be better to wait for @akien-mga to come back to give ok too, as this may break a lot of pending PRs if merged now.

**Comment by akien-mga** [comment]
Looks good to me, needs a rebase or redo, and fixing the build issues.

File: core/variant/binder_common.h [hunk 3]
```diff
@@ -78,7 +78,7 @@ struct VariantCaster<const T &> {
 		_FORCE_INLINE_ static m_enum convert(const void *p_ptr) {            \
 			return m_enum(*reinterpret_cast<const int64_t *>(p_ptr));        \
 		}                                                                    \
-		typedef int64_t EncodeT;                                             \
+		using EncodeT = int64_t;                                             \
 		_FORCE_INLINE_ static void encode(m_enum p_val, const void *p_ptr) { \
 			*(int64_t *)p_ptr = p_val;                                       \
 		}                                                                    \
```
Comment (aaronfranke) [inline core/variant/binder_common.h:84]: Is `EncodeT` dead code? It's not used before it goes out of scope.

Comment (akien-mga) [reply core/variant/binder_common.h:84]: No, it's used in `gdvirtual.gen.inc` generated by `core/object/make_virtuals.py`.

Comment (Calinou) [reply core/variant/binder_common.h:84]: I added a comment above the `#define` to clarify that EncodeT is being used in other files.

**Comment by akien-mga** [comment]
Still has some build issues for JS and iOS.

**Comment by Faless** [comment]
For the JavaScript part, those defines are inside the `extern "C" {}` declaration, so `using` is invalid.

**Comment by reduz** [comment]
I am not super convinced moving to a more modern C++ syntax for this makes a lot of sense. For many contributors, the typedef syntax is more understood as it comes from C. This is the same to me as moving enums used as costant to contexpr. Really not much of a point.

**Comment by fire** [comment]
Should we close this?

**Comment by Calinou** [comment]
There doesn't seem to be much interest in this, closing.

**Pull request closed without merging**